cmd/proxy/main_test.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2015 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+)
+
+func waitForProxy(t *testing.T, c *http.Client, apiUrl string) {
+	timeout := 5 * time.Second
+	deadline := time.Now().Add(timeout)
+	for time.Now().Before(deadline) {
+		if res, err := c.Get(apiUrl); err != nil || res.StatusCode != http.StatusOK {
+			time.Sleep(200 * time.Millisecond)
+			continue
+		}
+		return
+	}
+	t.Fatalf("waitForProxy: did not start up within %.1f seconds", timeout.Seconds())
+}
+
+// getFreePort returns a port string preceded by a colon, e.g. ":1234"
+func getFreePort(t *testing.T) string {
+	l, err := net.Listen("tcp", ":")
+	if err != nil {
+		t.Fatalf("getFreePort(): could not get free port: %v", err)
+	}
+	defer l.Close()
+	return l.Addr().String()[strings.LastIndex(l.Addr().String(), ":"):]
+}
+
+func TestProxyHttp(t *testing.T) {
+	tempDir, err := ioutil.TempDir("", t.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(tempDir)
+
+	// Build proxy binary
+	binPath := filepath.Join(tempDir, "proxy")
+	cmd := exec.Command("go", "build", "-o", binPath)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		t.Fatal(err)
+	}
+
+	// Start proxy
+	proxyPort := getFreePort(t)
```

Comment:
But people who run this test locally may not have those ports available, e.g. they already have one or two proxies or something else running that happen to use the same port as the ones we hardcoded in the test.  I don't see a downside to letting the system pick available ports.